scarb/src/ops/lint/mod.rs

```diff
@@ -0,0 +1,399 @@
+use cairo_lang_semantic::{
+    SemanticDiagnostic, db::SemanticGroup, diagnostic::SemanticDiagnosticKind,
+};
+use cairo_lint::{
+    CAIRO_LINT_TOOL_NAME, LinterAnalysisDatabase, LinterDiagnosticParams, LinterGroup,
+};
+use cairo_lint::{
+    CairoLintToolMetadata, apply_file_fixes, diagnostics::format_diagnostic, get_fixes,
+};
+use camino::Utf8PathBuf;
+use itertools::Itertools;
+use scarb_ui::components::Status;
+
+use crate::core::{Package, Workspace};
+use crate::internal::fsx::canonicalize;
+
+use super::{
+    CompilationUnitsOpts, FeaturesOpts, compile_unit, plugins_required_for_units, validate_features,
+};
+
+mod db;
+
+pub struct LintOptions {
+    pub packages: Vec<Package>,
+    pub target_names: Vec<String>,
+    pub test: bool,
+    pub fix: bool,
+    pub ignore_cairo_version: bool,
+    pub features: FeaturesOpts,
+    pub deny_warnings: bool,
+    pub path: Option<Utf8PathBuf>,
+}
+
+#[tracing::instrument(skip_all, level = "debug")]
+pub fn lint(opts: LintOptions, ws: &Workspace<'_>) -> Result<()> {
+    let resolve = ops::resolve_workspace(ws)?;
+
+    validate_features(&opts.packages, &opts.features)?;
+
+    let compilation_units = ops::generate_compilation_units(
+        &resolve,
+        &opts.features,
+        ws,
+        CompilationUnitsOpts {
+            ignore_cairo_version: opts.ignore_cairo_version,
+            load_prebuilt_macros: ws.config().load_prebuilt_proc_macros(),
+        },
+    )?;
+
+    let absolute_path = opts.path.map(canonicalize).transpose()?;
+
+    // Select proc macro units that need to be compiled for Cairo compilation units.
+    let required_plugins = plugins_required_for_units(&compilation_units);
+
+    // We process all proc-macro units that are required by Cairo compilation units beforehand.
+    for compilation_unit in compilation_units.iter() {
+        if let CompilationUnit::ProcMacro(_) = compilation_unit
+            && required_plugins.contains(&compilation_unit.main_package_id())
+        {
+            compile_unit(compilation_unit.clone(), ws)?;
+        }
+    }
+
+    let compilation_units_for_linting = get_compilation_units_for_linting(
+        ws,
+        &compilation_units,
+        &opts.packages,
+        opts.test,
+        &opts.target_names,
+    );
+
+    // We store the state of the workspace diagnostics, so we can decide upon throwing an error later on.
+    // Also we want to apply fixes only if there were no previous errors.
+    let mut packages_with_error: Vec<PackageName> = Default::default();
+
+    for (package, compilation_unit) in compilation_units_for_linting.iter() {
+        ws.config()
+            .ui()
+            .print(Status::new("Linting", &compilation_unit.name()));
+
+        let db = build_lint_database(compilation_unit, ws)?;
+        let diagnostics = get_linting_diagnostics(&db, package, &absolute_path, compilation_unit)?;
+        display_diagnostics(ws, &db, &diagnostics);
```

Comment:
this is going to display warnings even if `--no-warnings` flag is passed, is it on purpose (and if so what's the reasoning behind)?